src/Propulsion.MessageDb/MessageDbSource.fs

```diff
@@ -0,0 +1,97 @@
+        use conn = new NpgsqlConnection(connectionString)
+        do! conn.OpenAsync(ct)
+        let command = conn.CreateCommand()
+        command.CommandText <- "select
+                                   global_position, type, data, metadata, id::uuid,
+                                   (metadata::jsonb->>'$correlationId')::text,
+                                   (metadata::jsonb->>'$causationId')::text,
+                                   time, stream_name
+                                 from get_category_messages(@Category, @Position, @BatchSize);"
+        command.Parameters.AddWithValue("Category", NpgsqlDbType.Text, TrancheId.toString category) |> ignore
+        command.Parameters.AddWithValue("Position", NpgsqlDbType.Bigint, fromPositionInclusive) |> ignore
+        command.Parameters.AddWithValue("BatchSize", NpgsqlDbType.Bigint, int64 batchSize) |> ignore
+
+        let! reader = command.ExecuteReaderAsync(ct)
+        let! hasRow = reader.ReadAsync(ct)
+        let mutable hasRow = hasRow
+
+        let events = ResizeArray()
+        while hasRow do
+            events.Add(readRow reader)
+            let! nextHasRow = reader.ReadAsync(ct)
+            hasRow <- nextHasRow
+        return events.ToArray() }
+    member _.ReadCategoryLastVersion(category: TrancheId, ct) = task {
+        use conn = new NpgsqlConnection(connectionString)
+        do! conn.OpenAsync(ct)
+        let command = conn.CreateCommand()
+        command.CommandText <- "select max(global_position) from messages where category(stream_name) = @Category;"
+        command.Parameters.AddWithValue("Category", NpgsqlDbType.Text, TrancheId.toString category) |> ignore
+
+        let! reader = command.ExecuteReaderAsync(ct)
+        let! hasRow = reader.ReadAsync(ct)
+        return if hasRow then reader.GetInt64(0) else 0L
+    }
+
+module private Impl =
+    open Propulsion.Infrastructure // AwaitTaskCorrect
+
+    let readBatch batchSize (store : MessageDbCategoryReader) (category, pos) : Async<Propulsion.Feed.Core.Batch<_>> = async {
+        let! ct = Async.CancellationToken
+        let positionInclusive = Position.toInt64 pos
+        let! page = store.ReadCategoryMessages(category, positionInclusive, batchSize, ct) |> Async.AwaitTaskCorrect
+        let checkpoint = match Array.tryLast page with Some struct(_, evt) -> evt.Index | None -> positionInclusive
+        return { checkpoint = Position.parse checkpoint; items = page; isTail = false } }
+
+    let readTailPositionForTranche (store : MessageDbCategoryReader) trancheId : Async<Propulsion.Feed.Position> = async {
+        let! ct = Async.CancellationToken
+        let! lastEventPos = store.ReadCategoryLastVersion(trancheId, ct) |> Async.AwaitTaskCorrect
+        return Position.parse(lastEventPos) }
+
+type MessageDbSource
+    (   log : Serilog.ILogger, statsInterval,
+        reader: MessageDbCategoryReader, batchSize, tailSleepInterval,
```

Comment:
source should coordinate a set of readers, potentially across >1 category

DynamoStoreSource is wired for this sort of thing, see https://github.com/jet/propulsion/blob/master/src/Propulsion.DynamoStore/DynamoStoreSource.fs#L169-L184

As such, the ctor should take a list of category names (prob as a `string array`)
If mdb presents a category list, you could event implement the equivalent of the "loading the registered tranches list a la https://github.com/jet/propulsion/blob/master/src/Propulsion.DynamoStore/DynamoStoreSource.fs#L176

However in general a given handler knows the categories it wants, so the CreateSource wiring should pass that list
https://github.com/jet/dotnet-templates/blob/01af08355246f76ad292c714ab742003aa9605b9/equinox-shipping/Watchdog/Handler.fs#L31

Reply:
I'm not sure I follow exactly what you're saying.

> source should coordinate a set of readers, potentially across >1 category

Unless I've missed something this setup allows reading across categories already

```fsharp
use src = source.Start(source.Pump(fun _ -> async { 
  // pass as many categories in as you'd like
  return [| TrancheId.parse category1; TrancheId.parse category2 |] }))
```

It's likely that I'm missing context, and "reader" is too overloaded

> As such, the ctor should take a list of category names (prob as a string array)

As above, isn't this already taken care of by the Pump's ability to pick tranches?

> loading the registered tranches list 

This should be a fairly simple `select distinct category(stream_name) from messages` but is not a built in API.

Reply:
Oh wait a second... I get it now 🤦. I had taken the example from the memory store test and thought it represented a public API.

Reply:
Alright, added categories as a parameter now.

On the auto loading of tranches... I'm not sure "subscribing to all" is necessarily a good idea with message-db and at the very least, if it's something a user wants to do they can run a query like the one above and pass the result into the subscription.

Reply:
yes not saying subscribing to all should be a mainline feature - having to pass a static list is my preferred main API
The `propulsion` tool has a facility to walk all events (wiring that up is optional, I did it for completeness more as a way to sanity check perf of cosmosdb changefeed than anything really useful) and do a null projection on them, and the `proArchiver` template also needs to walk all, but that's very much an optional feature

Reply:
to match the other store sources, the arg should not be a Reader, but instead a connection string (it can be argued that the other stores should follow this shape, but the current impls might not be stateless)

Reply:
https://github.com/jet/propulsion/blob/master/src/Propulsion.DynamoStore/DynamoStoreSource.fs#L144

🤔 DynamoStoreSource was what I modeled after and it takes a `DynamoStoreClient` directly.

Reply:
Fair ;) I guess the thing that's bothering me is that it's a _Reader_
On the DynamoStore side, there's various layers to it going on as it's ultimately reading an Index that's actually Equinox events, so its necessarily touching Equinox low level components
It's a plus for Propulsion.Mdb and .Sss that they don't have a dependency on Equinox.
... ok, finally realising that Reader is referring to Reader vs Writer in the Equinox impl...
Perhaps if it's renamed to `MessageDbCategoryClient`
... (though I still think it makes sense to say that the Client is an implementation detail - Propulsion.MessageDb only needs a connection string in order to take care of it all. The counterpoint to that is that any additional configurability would benefit from a separated Client with e.g. extra properties or ctors.)
Looking more closely at SSS, it depends on `IStreamStore` (which I have just noticed can be the base interface of that, `IReadonlyStreamStore`). This does require a separate create client / connect phase.

... not seeing a single perfect answer, other than making sure there's no `Reader` to clash with the internal thing of the same name ;)

Reply:
I think a connstr API would be nicer to use as a consumer so I'd be happy to make it like that.

Can you help me with the magic syntax incantation?

```fsharp
type Thing(connstr)
  let reader = Reader(connstr)
  inherit OtherThing(reader)
```

Or were you expecting it to be created for each asyncSeq?

```fsharp
 (fun req -> asyncSeq {
                let sw = Stopwatch.StartNew()
                let! b = Impl.readBatch batchSize (Reader connectionString) req
                yield sw.Elapsed, b })
```

Reply:
In the meantime I've renamed to Client

Reply:
Dont think there is a syntax for it - if you do find one, DynamoStoreSource has about 4 places where it constructs a `DynamoStoreContext` from the client.
All you can do is have a chain of ctors, but that gets ugly fast when you have so many args.
I would not worry about a few constructs of the client in the greater scheme of things though

Reply:
Maybe you need an `op_implicit` :P

Reply:
> All you can do is have a chain of ctors, but that gets ugly fast when you have so many args.

I went for this and I think I'm okay with the result

Reply:
Maybe make the primary ctor and `MessageDbCategoryClient` internal to avoid having to arbitrarily select in intellisense?